### PR TITLE
feat: add filter for query in ts templates

### PR DESCRIPTION
**Pull request by thucpn**
SimpleVectorStore:
- generate -> private = false (public docs) -> store in data/
- /chat/upload -> private = true (private docs) -> store in output/uploaded/

LlamaCloudIndex:
- upload directly in LlamaCloud -> pipeline_id != null & private = undefined (public docs) -> store in data/llamacloud/
- generate -> private = false (public docs) -> store in data/
- /chat/upload -> private = true (private docs) -> store in output/uploaded/

## Summary by CodeRabbit

- **New Features**
  - Introduced a filtering mechanism for queries in TypeScript templates, enhancing data handling and usability.
  - Added a unique `doc_id` property to document metadata for better tracking.
  - Implemented a dynamic filter generation process in the chat engine for tailored queries.
  - Structured the Next.js configuration to utilize enhanced integration with LlamaIndex.
  - Enhanced filtering logic in the FastAPI chat router to accommodate multiple values.

- **Bug Fixes**
  - Updated document ID retrieval mechanism to accommodate new data structure expectations.

- **Chores**
  - Enhanced configuration file management for Next.js applications.
  - Removed the `file_metadata_func`, indicating a shift in file metadata handling.

### CODE REVIEW

File: templates/types/streaming/nextjs/next.config.mjs [hunk 1]
```diff
@@ -1,8 +1,9 @@
 /** @type {import('next').NextConfig} */
 import fs from "fs";
+import withLlamaIndex from "llamaindex/next";
 import webpack from "./webpack.config.mjs";
 
 const nextConfig = JSON.parse(fs.readFileSync("./next.config.json", "utf-8"));
 nextConfig.webpack = webpack;
 
-export default nextConfig;
+export default withLlamaIndex(nextConfig);
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/next.config.mjs:9]: i think you need to take care that the different `webpack.config.mjs` files are working

File: templates/components/llamaindex/typescript/streaming/service.ts [hunk 2]
```diff
@@ -1,86 +1,66 @@
...
-          const localFileUrl = this.downloadFile(fileDetail.url, fileId, name);
-          return localFileUrl;
-        }
+  private static nodesToDownloadFiles(nodes: NodeWithScore<Metadata>[]) {
+    const downloadFiles: Array<{
+      pipelineId: string;
+      fileName: string;
+    }> = [];
+    for (const node of nodes) {
+      const isLocalFile = node.node.metadata["is_local_file"] === "true";
```
Comment (marcusschiesser) [inline templates/components/llamaindex/typescript/streaming/service.ts:36]: This metadata is not llamacloud specific

File: templates/components/engines/typescript/agent/chat.ts [hunk 3]
```diff
@@ -41,3 +46,32 @@ export async function createChatEngine(documentIds?: string[]) {
...
+        },
+      ],
+    };
+  }
+  return {
+    filters: [
+      {
+        key: "private",
+        value: "true",
+        operator: "!=",
```
[the remark this answers is not in the document]
Comment (leehuwuj) [reply templates/components/engines/typescript/agent/chat.ts:67]: My bad! It should be `NIN` filter after the LlamaCloud feature.

File: templates/components/vectordbs/typescript/llamacloud/generate.ts [hunk 4]
```diff
@@ -9,12 +9,10 @@ dotenv.config();
 
 async function loadAndIndex() {
   const documents = await getDocuments();
-  // Set is_local_file=true to distinguish locally ingested files from LlamaCloud files
   // Set private=false to mark the document as public (required for filtering)
   for (const document of documents) {
     document.metadata = {
       ...document.metadata,
-      is_local_file: "true",
```
Comment (marcusschiesser) [inline templates/components/vectordbs/typescript/llamacloud/generate.ts:16]: @thucpn we still need `is_local_file` here to distinguish the files that run with generate script but end up in LlamaCloud

**Pull request merged**